overlord/snapstate/component.go

```diff
@@ -30,8 +31,175 @@ import (
+	for _, compsup := range compsups {
+		// since we are installing multiple components, we don't want to setup
+		// the security profiles until the end
+		compsup.SkipProfiles = true
+
+		componentTS, err := doInstallComponent(st, &snapst, compsup, snapsup, opts.FromChange)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(componentTS.beforeLink) == 0 && len(componentTS.linkToHook) == 0 {
+			return nil, fmt.Errorf("internal error: missing expected tasks in component task set")
+		}
+
+		// splice the setup security task into chain of tasks, right before we
+		// link the component. this means that all tasks for all components will
+		// run up until the link-component task. then one setup-security task
+		// will run, allowing all of the link-component tasks to run.
+		setupSecurity.WaitFor(componentTS.beforeLink[len(componentTS.beforeLink)-1])
+		componentTS.linkToHook[0].WaitFor(setupSecurity)
+
+		compSetupIDs = append(compSetupIDs, componentTS.compSetupTask.ID())
+		tss = append(tss, componentTS.taskSet())
+	}
+
+	setupSecurity.Set("component-setup-tasks", compSetupIDs)
+
+	tss = append(tss, state.NewTaskSet(setupSecurity))
+
+	return tss, nil
+}
+
+func componentSetupsForInstall(ctx context.Context, st *state.State, names []string, info *snap.Info, opts Options) ([]ComponentSetup, error) {
+	var snapst SnapState
+	err := Get(st, info.InstanceName(), &snapst)
+	if err != nil {
+		if errors.Is(err, state.ErrNoState) {
+			return nil, &snap.NotInstalledError{Snap: info.InstanceName()}
+		}
+		return nil, err
+	}
+
+	current, err := currentSnaps(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO:COMPS: figure out which user to use here
+	user, err := userFromUserID(st, opts.UserID)
+	if err != nil {
+		return nil, err
+	}
+
+	action, err := installComponentAction(st, snapst, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	refreshOpts, err := refreshOptions(st, &store.RefreshOptions{
+		IncludeResources: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	sto := Store(st, opts.DeviceCtx)
+	st.Unlock()
+	sars, _, err := sto.SnapAction(ctx, current, []*store.SnapAction{action}, nil, user, refreshOpts)
+	st.Lock()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(sars) != 1 {
+		return nil, fmt.Errorf("internal error: expected exactly one snap action result, got %d", len(sars))
+	}
+
+	return componentTargetsFromActionResult("install", sars[0], names)
+}
+
+func installComponentAction(st *state.State, snapst SnapState, opts Options) (*store.SnapAction, error) {
+	si := snapst.CurrentSideInfo()
+	if si == nil {
+		return nil, errors.New("internal error: cannot install components for a snap that is not installed")
+	}
+
+	if si.SnapID == "" {
+		return nil, errors.New("internal error: cannot install components for a snap that is unknown to the store")
+	}
+
+	enforcedSetsFunc := cachedEnforcedValidationSets(st)
+
+	// we send a refresh action, since that is what the store requested that
+	// we do in this case
```

Comment:
Which was the specific reason for this? Maybe worth expanding the comment for the rationale.